Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,257 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type `\forall`
+
+-- Lean treats the universal quantifier as an implication
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention for operations like +, *, ...)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+
+-- Lean has a special type for propositions, called `Prop`
+#check a+b=δ
+
+-- Lean doesn't care if a Proposition is true or false, as long as it is well-formed
+#check 2+2=5
+#check 1=1
+#check == -- Delete this line to get rid of the error message
+
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+
+-- An implication is like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+-- Notice the syntax: if `f` is a function, Lean reads `f x` as `f(x)` (and gives error if `x` is not a valid input)
+-- Lean automatically groups on the left, so `f x y` means `f(x)(y)`
+-- If you need a different order you need to specify it with appropriate brackets
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · exact h₀
+  · exact h₁
+
+
+-- Using the universal quantifier, we define an upper bounded function
+def FnUb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, f x ≤ a
+
+
+
+variable (f g : ℝ → ℝ) (a b : ℝ)
+
+-- It is useful to define functions, this can be done with the command `fun`
+
+#check fun x ↦ x + 1
+#check fun x ↦ f x + g x
+
+-- We can use implication hypothesis to get new hypotheses
+
+example (hfa : FnUb f a) (hgb : FnUb g b) : FnUb (fun x ↦ f x + g x) (a + b) := by
+  intro x
+  dsimp
+  -- This step makes Lean unpack the definition of the function.
+  -- It is not needed, but it makes things clearer
+  apply add_le_add
+  · exact hfa x
+  -- `hfa x` is now a proof of `f x ≤ a`
+  · exact hgb x
+
+/-
+Since Lean sees an implication as a function from proofs of a Proposition to
+proofs of another Proposition, we can write the proof term explicitly, using `fun`.
+This is more concise, but probably less "human-like." Where to draw the line between
+conciseness and human readability is up to you!
+-/
+
+theorem fnUb_add {f g : ℝ → ℝ} {a b : ℝ} (hfa : FnUb f a) (hgb : FnUb g b) :
```

Comment:
You're right, perhaps we should also say something about hypotheses of theorems in general?
Like they are logically just an implication, but it's convenient to keep them has hypotheses because Lean allows for the curly bracket and hence implicit variables (which they should know from last session)

Reply:
I actually don't think we covered implicit variables before, so we should either explain them here or keep them explicit and cover implicit variables another time.

Reply:
We should explain how the named hypotheses work. For example, we could say that
```
variable (P Q : Prop)

example : P → Q := by
  intro hp
  ...
```
is the same as
```
example (hp : P) : Q := by
  ...
```

Reply:
I like @AdrianDoM's suggestion